coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/UnionSqlCallTransformer.java

```diff
@@ -0,0 +1,165 @@
+
+      List<SqlNode> rewrittenSelectNodes = new ArrayList<>();
+      boolean useRewrittenSelectNodes = false;
+      for (int i = 0; i < selectNodes.size(); i++) {
+
+        if (columnsTypesLists.get(i).stream().anyMatch(columnType -> !columnType.isPresent())) {
+          // Couldn't determine the type for all the expressions
+          continue;
+        }
+        RelDataType inferredColumnType =
+            leastRestrictive(columnsTypesLists.get(i).stream().map(Optional::get).collect(Collectors.toList()));
+
+        SqlNode selectNode = selectNodes.get(i);
+        RelDataType selectNodeDataType = deriveRelDatatype(selectNodes.get(i));
+        if (!selectNodeDataType.equals(inferredColumnType) && selectNodeDataType.getSqlTypeName() == SqlTypeName.CHAR
+            && inferredColumnType.getSqlTypeName() == SqlTypeName.VARCHAR) {
+          // Work-around for the Trino limitation in dealing UNION statements between `char` and `varchar`.
+          // See https://github.com/trinodb/trino/issues/9031
+          SqlNode rewrittenSelectNode = castNode(selectNode, inferredColumnType);
+          if (!useRewrittenSelectNodes) {
+            rewrittenSelectNodes.addAll(selectNodes.subList(0, i));
+            useRewrittenSelectNodes = true;
+          }
+          rewrittenSelectNodes.add(rewrittenSelectNode);
+        } else if (useRewrittenSelectNodes) {
+          rewrittenSelectNodes.add(selectNode);
+        }
+      }
+
+      if (useRewrittenSelectNodes) {
+        select.setSelectList(new SqlNodeList(rewrittenSelectNodes, SqlParserPos.ZERO));
+      }
+      updatedOperands.add(operand);
+    }
+    return sqlCall.getOperator().createCall(POS, updatedOperands);
+  }
+
+  private SqlNode castNode(SqlNode node, RelDataType type) {
+    if (node.getKind() == SqlKind.AS) {
+      SqlNode expression = ((SqlCall) node).getOperandList().get(0);
+      SqlIdentifier identifier = (SqlIdentifier) ((SqlCall) node).getOperandList().get(1);
+      return SqlStdOperatorTable.AS.createCall(POS, new SqlCastFunction() {
+        @Override
+        public RelDataType deriveType(SqlValidator validator, SqlValidatorScope scope, SqlCall call) {
+          SqlCallBinding opBinding = new SqlCallBinding(validator, scope, call);
+          return inferReturnType(opBinding);
+        }
+      }.createCall(ZERO, expression, getSqlDataTypeSpecForCasting(type)), identifier);
+    } else {
+      // If there's no existing alias, just do the cast
+      return new SqlCastFunction() {
+        @Override
+        public RelDataType deriveType(SqlValidator validator, SqlValidatorScope scope, SqlCall call) {
+          SqlCallBinding opBinding = new SqlCallBinding(validator, scope, call);
+          return inferReturnType(opBinding);
+        }
+      }.createCall(ZERO, node, getSqlDataTypeSpecForCasting(type));
+    }
+  }
+
+  private static SqlDataTypeSpec getSqlDataTypeSpecForCasting(RelDataType relDataType) {
```

Comment:
Seems there is no need for the `ForCasting` part since this is not specific to casing?